// @nx/angular/src/builders/utilities/module-federation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateDevRemotes = exports.getStaticRemotes = exports.getDynamicRemotes = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
const devkit_1 = require("@nx/devkit");
const internal_1 = require("@nx/js/src/internal");
function getDynamicRemotes(project, context, workspaceProjects, remotesToSkip, pathToManifestFile = (0, path_1.join)(context.workspaceRoot, project.sourceRoot, 'assets/module-federation.manifest.json')) {
    // check for dynamic remotes
    // we should only check for dynamic based on what we generate
    // and fallback to empty array
    if (!(0, fs_1.existsSync)(pathToManifestFile)) {
        return [];
    }
    const moduleFederationManifestJson = (0, fs_1.readFileSync)(pathToManifestFile, 'utf-8');
    if (!moduleFederationManifestJson) {
        return [];
    }
    // This should have shape of
    // {
    //   "remoteName": "remoteLocation",
    // }
    const parsedManifest = JSON.parse(moduleFederationManifestJson);
    if (!Object.keys(parsedManifest).every((key) => typeof key === 'string' && typeof parsedManifest[key] === 'string')) {
        return [];
    }
    const allDynamicRemotes = Object.entries(parsedManifest)
        .map(([remoteName]) => remoteName)
        .filter((r) => !remotesToSkip.has(r));
    const remotesNotInWorkspace = [];
    const dynamicRemotes = allDynamicRemotes.filter((remote) => {
        if (!workspaceProjects[remote]) {
            remotesNotInWorkspace.push(remote);
            return false;
        }
        return true;
    });
    if (remotesNotInWorkspace.length > 0) {
        devkit_1.logger.warn(`Skipping serving ${remotesNotInWorkspace.join(', ')} as they could not be found in the workspace. Ensure they are served correctly.`);
    }
    return dynamicRemotes;
}
exports.getDynamicRemotes = getDynamicRemotes;
function getModuleFederationConfig(tsconfigPath, workspaceRoot, projectRoot) {
    const moduleFederationConfigPathJS = (0, path_1.join)(workspaceRoot, projectRoot, 'module-federation.config.js');
    const moduleFederationConfigPathTS = (0, path_1.join)(workspaceRoot, projectRoot, 'module-federation.config.ts');
    let moduleFederationConfigPath = moduleFederationConfigPathJS;
    let cleanupTranspiler = () => { };
    if ((0, fs_1.existsSync)(moduleFederationConfigPathTS)) {
        cleanupTranspiler = (0, internal_1.registerTsProject)((0, path_1.join)(workspaceRoot, tsconfigPath));
        moduleFederationConfigPath = moduleFederationConfigPathTS;
    }
    try {
        const config = require(moduleFederationConfigPath);
        cleanupTranspiler();
        return {
            mfeConfig: config.default || config,
            mfConfigPath: moduleFederationConfigPath,
        };
    }
    catch {
        throw new Error(`Could not load ${moduleFederationConfigPath}. Was this project generated with "@nx/angular:host"?`);
    }
}
function getStaticRemotes(project, context, workspaceProjects, remotesToSkip) {
    const { mfeConfig, mfConfigPath } = getModuleFederationConfig(project.targets.build.options.tsConfig, context.workspaceRoot, project.root);
    const remotesConfig = Array.isArray(mfeConfig.remotes) && mfeConfig.remotes.length > 0
        ? mfeConfig.remotes
        : [];
    const allStaticRemotes = remotesConfig
        .map((remoteDefinition) => Array.isArray(remoteDefinition) ? remoteDefinition[0] : remoteDefinition)
        .filter((r) => !remotesToSkip.has(r));
    const remotesNotInWorkspace = [];
    const staticRemotes = allStaticRemotes.filter((remote) => {
        if (!workspaceProjects[remote]) {
            remotesNotInWorkspace.push(remote);
            return false;
        }
        return true;
    });
    if (remotesNotInWorkspace.length > 0) {
        devkit_1.logger.warn(`Skipping serving ${remotesNotInWorkspace.join(', ')} as they could not be found in the workspace. Ensure they are served correctly.`);
    }
    return staticRemotes;
}
exports.getStaticRemotes = getStaticRemotes;
function validateDevRemotes(options, workspaceProjects) {
    const invalidDevRemotes = options.devRemotes?.filter((remote) => !workspaceProjects[remote]) ?? [];
    if (invalidDevRemotes.length) {
        throw new Error(invalidDevRemotes.length === 1
            ? `Invalid dev remote provided: ${invalidDevRemotes[0]}.`
            : `Invalid dev remotes provided: ${invalidDevRemotes.join(', ')}.`);
    }
}
exports.validateDevRemotes = validateDevRemotes;
