// @nx/angular/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsoncEslintParserVersion = exports.jasmineMarblesVersion = exports.typesNodeVersion = exports.jestPresetAngularVersion = exports.tsNodeVersion = exports.autoprefixerVersion = exports.postcssUrlVersion = exports.postcssVersion = exports.tailwindVersion = exports.angularEslintVersion = exports.moduleFederationNodeVersion = exports.browserSyncVersion = exports.typesExpressVersion = exports.expressVersion = exports.typesCorsVersion = exports.corsVersion = exports.tsLibVersion = exports.angularJsVersion = exports.zoneJsVersion = exports.rxjsVersion = exports.ngrxVersion = exports.ngPackagrVersion = exports.angularDevkitVersion = exports.angularVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.angularVersion = '~17.3.0';
exports.angularDevkitVersion = '~17.3.0';
exports.ngPackagrVersion = '~17.3.0';
exports.ngrxVersion = '~17.0.0';
exports.rxjsVersion = '~7.8.0';
exports.zoneJsVersion = '~0.14.3';
exports.angularJsVersion = '1.7.9';
exports.tsLibVersion = '^2.3.0';
exports.corsVersion = '~2.8.5';
exports.typesCorsVersion = '~2.8.5';
exports.expressVersion = '~4.18.2';
exports.typesExpressVersion = '4.17.14';
exports.browserSyncVersion = '^3.0.0';
exports.moduleFederationNodeVersion = '~1.0.5';
exports.angularEslintVersion = '~17.3.0';
exports.tailwindVersion = '^3.0.2';
exports.postcssVersion = '^8.4.5';
exports.postcssUrlVersion = '~10.1.3';
exports.autoprefixerVersion = '^10.4.0';
exports.tsNodeVersion = '10.9.1';
exports.jestPresetAngularVersion = '~14.0.3';
exports.typesNodeVersion = '18.16.9';
exports.jasmineMarblesVersion = '^0.9.2';
exports.jsoncEslintParserVersion = '^2.1.0';
