// @nx/angular/src/generators/library/library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.libraryGeneratorInternal = exports.libraryGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const eslint_1 = require("@nx/eslint");
const js_1 = require("@nx/js");
const init_1 = require("../../generators/init/init");
const add_linting_1 = require("../add-linting/add-linting");
const setup_tailwind_1 = require("../setup-tailwind/setup-tailwind");
const version_utils_1 = require("../utils/version-utils");
const dependencies_1 = require("../utils/dependencies");
const add_module_1 = require("./lib/add-module");
const add_standalone_component_1 = require("./lib/add-standalone-component");
const enable_strict_type_checking_1 = require("./lib/enable-strict-type-checking");
const normalize_options_1 = require("./lib/normalize-options");
const update_lib_package_npm_scope_1 = require("./lib/update-lib-package-npm-scope");
const update_tsconfig_1 = require("./lib/update-tsconfig");
const create_files_1 = require("./lib/create-files");
const add_project_1 = require("./lib/add-project");
const add_jest_1 = require("../utils/add-jest");
const set_generator_defaults_1 = require("./lib/set-generator-defaults");
const ensure_angular_dependencies_1 = require("../utils/ensure-angular-dependencies");
const log_show_project_command_1 = require("@nx/devkit/src/utils/log-show-project-command");
async function libraryGenerator(tree, schema) {
    return await libraryGeneratorInternal(tree, {
        // provide a default projectNameAndRootFormat to avoid breaking changes
        // to external generators invoking this one
        projectNameAndRootFormat: 'derived',
        ...schema,
    });
}
exports.libraryGenerator = libraryGenerator;
async function libraryGeneratorInternal(tree, schema) {
    // Do some validation checks
    if (!schema.routing && schema.lazy) {
        throw new Error(`To use "--lazy" option, "--routing" must also be set.`);
    }
    if (schema.publishable === true && !schema.importPath) {
        throw new Error(`For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`);
    }
    if (schema.addTailwind && !schema.buildable && !schema.publishable) {
        throw new Error(`To use "--addTailwind" option, you have to set either "--buildable" or "--publishable".`);
    }
    const options = await (0, normalize_options_1.normalizeOptions)(tree, schema);
    const { libraryOptions } = options;
    const pkgVersions = (0, version_utils_1.versions)(tree);
    await (0, js_1.initGenerator)(tree, {
        ...libraryOptions,
        js: false,
        skipFormat: true,
    });
    await (0, init_1.default)(tree, { ...libraryOptions, skipFormat: true });
    (0, ensure_angular_dependencies_1.ensureAngularDependencies)(tree);
    const project = (0, add_project_1.addProject)(tree, libraryOptions);
    (0, create_files_1.createFiles)(tree, options, project);
    (0, update_tsconfig_1.updateTsConfig)(tree, libraryOptions);
    await addUnitTestRunner(tree, libraryOptions);
    updateNpmScopeIfBuildableOrPublishable(tree, libraryOptions);
    (0, set_generator_defaults_1.setGeneratorDefaults)(tree, options);
    if (!libraryOptions.standalone) {
        (0, add_module_1.addModule)(tree, libraryOptions);
    }
    else {
        await (0, add_standalone_component_1.addStandaloneComponent)(tree, options);
    }
    setStrictMode(tree, libraryOptions);
    await addLinting(tree, libraryOptions);
    if (libraryOptions.addTailwind) {
        await (0, setup_tailwind_1.default)(tree, {
            project: libraryOptions.name,
            skipFormat: true,
            skipPackageJson: libraryOptions.skipPackageJson,
        });
    }
    if (libraryOptions.buildable || libraryOptions.publishable) {
        (0, version_utils_1.addDependenciesToPackageJsonIfDontExist)(tree, {}, {
            'ng-packagr': pkgVersions.ngPackagrVersion,
        });
        (0, dependencies_1.addBuildableLibrariesPostCssDependencies)(tree);
    }
    (0, js_1.addTsConfigPath)(tree, libraryOptions.importPath, [
        (0, devkit_1.joinPathFragments)(libraryOptions.projectRoot, './src', 'index.ts'),
    ]);
    if (!libraryOptions.skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
    return () => {
        (0, devkit_1.installPackagesTask)(tree);
        (0, log_show_project_command_1.logShowProjectCommand)(libraryOptions.name);
    };
}
exports.libraryGeneratorInternal = libraryGeneratorInternal;
async function addUnitTestRunner(host, options) {
    if (options.unitTestRunner === 'jest') {
        await (0, add_jest_1.addJest)(host, {
            name: options.name,
            projectRoot: options.projectRoot,
            skipPackageJson: options.skipPackageJson,
            strict: options.strict,
        });
    }
}
function updateNpmScopeIfBuildableOrPublishable(host, options) {
    if (options.buildable || options.publishable) {
        (0, update_lib_package_npm_scope_1.updateLibPackageNpmScope)(host, options);
    }
}
function setStrictMode(host, options) {
    if (options.strict) {
        (0, enable_strict_type_checking_1.enableStrictTypeChecking)(host, options);
    }
    else {
        (0, enable_strict_type_checking_1.setLibraryStrictDefault)(host, options.strict);
    }
}
async function addLinting(host, options) {
    if (options.linter === eslint_1.Linter.None) {
        return;
    }
    await (0, add_linting_1.default)(host, {
        projectName: options.name,
        projectRoot: options.projectRoot,
        prefix: options.prefix,
        unitTestRunner: options.unitTestRunner,
        setParserOptionsProject: options.setParserOptionsProject,
        skipFormat: true,
        skipPackageJson: options.skipPackageJson,
    });
}
exports.default = libraryGenerator;
