// @nx/angular/src/utils/mf/with-module-federation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withModuleFederation = void 0;
const utils_1 = require("./utils");
const ModuleFederationPlugin = require("webpack/lib/container/ModuleFederationPlugin");
async function withModuleFederation(options) {
    if (global.NX_GRAPH_CREATION) {
        return (config) => config;
    }
    const { sharedLibraries, sharedDependencies, mappedRemotes } = await (0, utils_1.getModuleFederationConfig)(options);
    return (config) => ({
        ...(config ?? {}),
        output: {
            ...(config.output ?? {}),
            uniqueName: options.name,
            publicPath: 'auto',
        },
        optimization: {
            ...(config.optimization ?? {}),
            runtimeChunk: false,
        },
        resolve: {
            ...(config.resolve ?? {}),
            alias: {
                ...(config.resolve?.alias ?? {}),
                ...sharedLibraries.getAliases(),
            },
        },
        experiments: {
            ...(config.experiments ?? {}),
            outputModule: true,
        },
        plugins: [
            ...(config.plugins ?? []),
            new ModuleFederationPlugin({
                name: options.name,
                filename: 'remoteEntry.mjs',
                exposes: options.exposes,
                remotes: mappedRemotes,
                shared: {
                    ...sharedDependencies,
                },
                library: {
                    type: 'module',
                },
            }),
            sharedLibraries.getReplacementPlugin(),
        ],
    });
}
exports.withModuleFederation = withModuleFederation;
