// @nx/angular/src/utils/mf/with-module-federation-ssr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withModuleFederationForSSR = void 0;
const utils_1 = require("./utils");
async function withModuleFederationForSSR(options) {
    if (global.NX_GRAPH_CREATION) {
        return (config) => config;
    }
    const { sharedLibraries, sharedDependencies, mappedRemotes } = await (0, utils_1.getModuleFederationConfig)(options, {
        isServer: true,
    });
    return (config) => ({
        ...(config ?? {}),
        target: false,
        output: {
            ...(config.output ?? {}),
            uniqueName: options.name,
        },
        optimization: {
            ...(config.optimization ?? {}),
            runtimeChunk: false,
        },
        resolve: {
            ...(config.resolve ?? {}),
            alias: {
                ...(config.resolve?.alias ?? {}),
                ...sharedLibraries.getAliases(),
            },
        },
        plugins: [
            ...(config.plugins ?? []),
            new (require('@module-federation/node').UniversalFederationPlugin)({
                name: options.name,
                filename: 'remoteEntry.js',
                exposes: options.exposes,
                remotes: mappedRemotes,
                shared: {
                    ...sharedDependencies,
                },
                library: {
                    type: 'commonjs-module',
                },
                isServer: true,
            }, {}),
            sharedLibraries.getReplacementPlugin(),
        ],
    });
}
exports.withModuleFederationForSSR = withModuleFederationForSSR;
