// @nx/workspace/src/generators/ci-workflow/ci-workflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ciWorkflowGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const default_base_1 = require("../../utilities/default-base");
const path_1 = require("path");
const nx_cloud_utils_1 = require("nx/src/utils/nx-cloud-utils");
async function ciWorkflowGenerator(tree, schema) {
    const ci = schema.ci;
    const nxJson = (0, devkit_1.readJson)(tree, 'nx.json');
    const nxCloudUsed = nxJson.nxCloudAccessToken ??
        Object.values(nxJson.tasksRunnerOptions ?? {}).find((r) => r.runner == '@nrwl/nx-cloud' || r.runner == 'nx-cloud');
    if (!nxCloudUsed) {
        throw new Error('This workspace is not connected to Nx Cloud.');
    }
    if (ci === 'bitbucket-pipelines' && defaultBranchNeedsOriginPrefix(nxJson)) {
        (0, devkit_1.writeJson)(tree, 'nx.json', appendOriginPrefix(nxJson));
    }
    const options = normalizeOptions(schema, tree);
    (0, devkit_1.generateFiles)(tree, (0, path_1.join)(__dirname, 'files', ci), '', options);
    await (0, devkit_1.formatFiles)(tree);
}
exports.ciWorkflowGenerator = ciWorkflowGenerator;
function normalizeOptions(options, tree) {
    const { name: workflowName, fileName: workflowFileName } = (0, devkit_1.names)(options.name);
    const packageManager = (0, devkit_1.detectPackageManager)();
    const { exec: packageManagerPrefix, ciInstall: packageManagerInstall } = (0, devkit_1.getPackageManagerCommand)(packageManager);
    const nxCloudUrl = (0, nx_cloud_utils_1.getNxCloudUrl)((0, devkit_1.readJson)(tree, 'nx.json'));
    const nxCloudHost = new URL(nxCloudUrl).host;
    const packageJson = (0, devkit_1.readJson)(tree, 'package.json');
    const allDependencies = {
        ...packageJson.dependencies,
        ...packageJson.devDependencies,
    };
    const hasE2E = allDependencies['@nx/cypress'] || allDependencies['@nx/playwright'];
    return {
        workflowName,
        workflowFileName,
        packageManager,
        packageManagerInstall,
        packageManagerPrefix,
        mainBranch: (0, default_base_1.deduceDefaultBase)(),
        hasE2E,
        nxCloudHost,
        tmpl: '',
    };
}
function defaultBranchNeedsOriginPrefix(nxJson) {
    const base = nxJson.defaultBase ?? nxJson.affected?.defaultBase;
    return !base?.startsWith('origin/');
}
function appendOriginPrefix(nxJson) {
    if (nxJson?.affected?.defaultBase) {
        nxJson.affected.defaultBase = `origin/${nxJson.affected.defaultBase}`;
    }
    if (nxJson.defaultBase || !nxJson.affected) {
        nxJson.defaultBase = `origin/${nxJson.defaultBase ?? (0, default_base_1.deduceDefaultBase)()}`;
    }
    return nxJson;
}
