// @nx/angular/src/generators/library/lib/add-standalone-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addStandaloneComponent = void 0;
const devkit_1 = require("@nx/devkit");
const component_1 = require("../../component/component");
const add_children_1 = require("./add-children");
const add_load_children_1 = require("./add-load-children");
async function addStandaloneComponent(tree, { libraryOptions, componentOptions }) {
    await (0, component_1.componentGenerator)(tree, {
        ...componentOptions,
        name: componentOptions.name,
        directory: (0, devkit_1.joinPathFragments)(libraryOptions.projectRoot, 'src', 'lib', componentOptions.flat ? '' : componentOptions.name),
        nameAndDirectoryFormat: 'as-provided',
        standalone: true,
        export: true,
        skipFormat: true,
    });
    if (libraryOptions.routing) {
        if (libraryOptions.parent) {
            if (libraryOptions.lazy) {
                (0, add_load_children_1.addLoadChildren)(tree, libraryOptions);
            }
            else {
                (0, add_children_1.addChildren)(tree, libraryOptions);
            }
        }
    }
}
exports.addStandaloneComponent = addStandaloneComponent;
