// @nx/angular/src/generators/library/lib/add-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addProject = void 0;
const devkit_1 = require("@nx/devkit");
const add_build_target_defaults_1 = require("@nx/devkit/src/generators/add-build-target-defaults");
function addProject(tree, libraryOptions) {
    const project = {
        name: libraryOptions.name,
        root: libraryOptions.projectRoot,
        sourceRoot: (0, devkit_1.joinPathFragments)(libraryOptions.projectRoot, 'src'),
        prefix: libraryOptions.prefix,
        tags: libraryOptions.parsedTags,
        projectType: 'library',
        targets: {},
    };
    if (libraryOptions.buildable || libraryOptions.publishable) {
        const executor = libraryOptions.publishable
            ? '@nx/angular:package'
            : '@nx/angular:ng-packagr-lite';
        (0, add_build_target_defaults_1.addBuildTargetDefaults)(tree, executor);
        project.targets.build = {
            executor,
            outputs: ['{workspaceRoot}/dist/{projectRoot}'],
            options: {
                project: `${libraryOptions.projectRoot}/ng-package.json`,
            },
            configurations: {
                production: {
                    tsConfig: `${libraryOptions.projectRoot}/tsconfig.lib.prod.json`,
                },
                development: {
                    tsConfig: `${libraryOptions.projectRoot}/tsconfig.lib.json`,
                },
            },
            defaultConfiguration: 'production',
        };
    }
    (0, devkit_1.addProjectConfiguration)(tree, libraryOptions.name, project);
    return project;
}
exports.addProject = addProject;
