// @nx/angular/src/generators/library/lib/enable-strict-type-checking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLibraryStrictDefault = exports.enableStrictTypeChecking = void 0;
const devkit_1 = require("@nx/devkit");
/**
 * Enable Strict Mode in the library and spec TS Config
 * */
function enableStrictTypeChecking(host, options) {
    updateTsConfig(host, options);
}
exports.enableStrictTypeChecking = enableStrictTypeChecking;
function setLibraryStrictDefault(host, isStrict) {
    // set the default so future libraries use it
    // unless the user has previously set this value
    const nxJson = (0, devkit_1.readNxJson)(host);
    nxJson.generators = nxJson.generators || {};
    nxJson.generators['@nx/angular:library'] =
        nxJson.generators['@nx/angular:library'] || {};
    nxJson.generators['@nx/angular:library'].strict =
        nxJson.generators['@nx/angular:library'].strict ?? isStrict;
    (0, devkit_1.updateNxJson)(host, nxJson);
}
exports.setLibraryStrictDefault = setLibraryStrictDefault;
function updateTsConfig(host, options) {
    // Update the settings in the tsconfig.app.json to enable strict type checking.
    // This matches the settings defined by the Angular CL https://angular.io/guide/strict-mode
    (0, devkit_1.updateJson)(host, `${options.projectRoot}/tsconfig.json`, (json) => {
        // update the TypeScript settings
        json.compilerOptions = {
            ...(json.compilerOptions ?? {}),
            forceConsistentCasingInFileNames: true,
            strict: true,
            noImplicitOverride: true,
            noPropertyAccessFromIndexSignature: true,
            noImplicitReturns: true,
            noFallthroughCasesInSwitch: true,
        };
        // update Angular Template Settings
        json.angularCompilerOptions = {
            ...(json.angularCompilerOptions ?? {}),
            enableI18nLegacyMessageIdFormat: false,
            strictInjectionParameters: true,
            strictInputAccessModifiers: true,
            strictTemplates: true,
        };
        return json;
    });
}
