// @nx/angular/src/generators/storybook-configuration/lib/generate-stories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateStories = void 0;
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../../utils/versions");
const stories_1 = require("../../stories/stories");
async function generateStories(tree, options) {
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    (0, devkit_1.ensurePackage)('@nx/cypress', versions_1.nxVersion);
    const { getE2eProjectName } = await Promise.resolve().then(() => require('@nx/cypress/src/utils/project-name'));
    const e2eProjectName = getE2eProjectName(options.project, project.root, options.cypressDirectory);
    await (0, stories_1.angularStoriesGenerator)(tree, {
        name: options.project,
        generateCypressSpecs: options.configureCypress && options.generateCypressSpecs,
        cypressProject: e2eProjectName,
        ignorePaths: options.ignorePaths,
        interactionTests: options.interactionTests,
        skipFormat: true,
    });
}
exports.generateStories = generateStories;
