// @nx/angular/src/generators/application/lib/create-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFiles = void 0;
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const update_project_root_tsconfig_1 = require("../../utils/update-project-root-tsconfig");
const test_runners_1 = require("../../../utils/test-runners");
const version_utils_1 = require("../../utils/version-utils");
const selector_1 = require("../../utils/selector");
async function createFiles(tree, options, rootOffset) {
    const { major: angularMajorVersion } = (0, version_utils_1.getInstalledAngularVersionInfo)(tree);
    const isUsingApplicationBuilder = angularMajorVersion >= 17 && options.bundler === 'esbuild';
    const rootSelector = `${options.prefix}-root`;
    (0, selector_1.validateHtmlSelector)(rootSelector);
    const nxWelcomeSelector = `${options.prefix}-nx-welcome`;
    (0, selector_1.validateHtmlSelector)(nxWelcomeSelector);
    const substitutions = {
        rootSelector,
        appName: options.name,
        inlineStyle: options.inlineStyle,
        inlineTemplate: options.inlineTemplate,
        style: options.style,
        viewEncapsulation: options.viewEncapsulation,
        unitTesting: options.unitTestRunner !== test_runners_1.UnitTestRunner.None,
        routing: options.routing,
        minimal: options.minimal,
        nxWelcomeSelector,
        rootTsConfig: (0, devkit_1.joinPathFragments)(rootOffset, (0, js_1.getRootTsConfigFileName)(tree)),
        angularMajorVersion,
        rootOffset,
        isUsingApplicationBuilder,
        tpl: '',
    };
    (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/base'), options.appProjectRoot, substitutions);
    if (options.standalone) {
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/standalone-components'), options.appProjectRoot, substitutions);
    }
    else {
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/ng-module'), options.appProjectRoot, substitutions);
    }
    (0, update_project_root_tsconfig_1.updateProjectRootTsConfig)(tree, options.appProjectRoot, (0, js_1.getRelativePathToRootTsConfig)(tree, options.appProjectRoot), options.rootProject);
    if (!options.routing) {
        tree.delete((0, devkit_1.joinPathFragments)(options.appProjectRoot, '/src/app/app.routes.ts'));
    }
    if (options.skipTests || options.unitTestRunner === test_runners_1.UnitTestRunner.None) {
        tree.delete((0, devkit_1.joinPathFragments)(options.appProjectRoot, '/src/app/app.component.spec.ts'));
    }
    if (options.inlineTemplate) {
        tree.delete((0, devkit_1.joinPathFragments)(options.appProjectRoot, '/src/app/app.component.html'));
    }
    if (options.inlineStyle) {
        tree.delete((0, devkit_1.joinPathFragments)(options.appProjectRoot, `/src/app/app.component.${options.style}`));
    }
    if (options.minimal) {
        tree.delete((0, devkit_1.joinPathFragments)(options.appProjectRoot, 'src/app/nx-welcome.component.ts'));
    }
}
exports.createFiles = createFiles;
