// @nx/angular/src/generators/setup-ssr/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const ast_utils_1 = require("../../../utils/nx-devkit/ast-utils");
const version_utils_1 = require("../../utils/version-utils");
function normalizeOptions(tree, options) {
    const isStandaloneApp = (0, ast_utils_1.isNgStandaloneApp)(tree, options.project);
    let hydration = options.hydration;
    if (hydration === undefined) {
        const { major: angularMajorVersion } = (0, version_utils_1.getInstalledAngularVersionInfo)(tree);
        hydration = angularMajorVersion >= 17;
    }
    return {
        project: options.project,
        appId: options.appId ?? 'serverApp',
        main: options.main ?? 'main.server.ts',
        serverFileName: options.serverFileName ?? 'server.ts',
        serverPort: options.serverPort ?? 4000,
        rootModuleFileName: options.rootModuleFileName ?? 'app.server.module.ts',
        rootModuleClassName: options.rootModuleClassName ?? 'AppServerModule',
        skipFormat: options.skipFormat ?? false,
        standalone: options.standalone ?? isStandaloneApp,
        hydration,
    };
}
exports.normalizeOptions = normalizeOptions;
