// @nx/angular/src/generators/setup-mf/lib/setup-serve-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupServeTarget = void 0;
const devkit_1 = require("@nx/devkit");
function setupServeTarget(host, options) {
    const appConfig = (0, devkit_1.readProjectConfiguration)(host, options.appName);
    appConfig.targets['serve'] = {
        ...appConfig.targets['serve'],
        executor: options.mfType === 'host'
            ? '@nx/angular:module-federation-dev-server'
            : '@nx/angular:dev-server',
        options: {
            ...appConfig.targets['serve'].options,
            port: options.port ?? undefined,
            publicHost: `http://localhost:${options.port ?? 4200}`,
        },
    };
    if (options.mfType === 'remote') {
        appConfig.targets['serve-static'] = {
            executor: '@nx/web:file-server',
            defaultConfiguration: 'production',
            options: {
                buildTarget: `${options.appName}:build`,
                port: options.port,
                watch: false,
            },
            configurations: {
                development: {
                    buildTarget: `${options.appName}:build:development`,
                },
                production: {
                    buildTarget: `${options.appName}:build:production`,
                },
            },
        };
    }
    (0, devkit_1.updateProjectConfiguration)(host, options.appName, appConfig);
}
exports.setupServeTarget = setupServeTarget;
